### PR TITLE
dist: Add GPG identity util

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: dist: Add GPG identity util
Additional Description:

Provides a lookup/discovery for signing keys required by #17289 

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

File: tools/gpg/identity.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+import gnupg
+
+
+class GPGError(Exception):
+    pass
+
+
+class GPGIdentity(object):
+    """A GPG identity and a signing key"""
```
Comment (phlax) [inline tools/gpg/identity.py:16]: add explanation about key lookup/matching

**Comment by phlax** [comment]
coverage is here https://storage.googleapis.com/envoy-pr/17399/tooling/index.html

File: tools/gpg/identity.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+        return os.path.join(self.home, ".gnupg")
+
+    @cached_property
+    def home(self) -> str:
+        """Gets *and sets if required* the `HOME` env var"""
+        os.environ["HOME"] = os.environ.get("HOME", pwd.getpwuid(os.getuid()).pw_dir)
+        return os.environ["HOME"]
+
+    @property
+    def log(self) -> str:
```
Comment (phlax) [inline tools/gpg/identity.py:59]: `logging.getLogger`

also return the default logger if its not set

File: tools/gpg/identity.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+
+    @cached_property
+    def home(self) -> str:
+        """Gets *and sets if required* the `HOME` env var"""
+        os.environ["HOME"] = os.environ.get("HOME", pwd.getpwuid(os.getuid()).pw_dir)
+        return os.environ["HOME"]
+
+    @cached_property
+    def log(self) -> logging.Logger:
+        return self._log or logging.getLogger()
```
Comment (phlax) [inline tools/gpg/identity.py:60]: `logging.getLogger(self.__name__)`

File: tools/gpg/identity.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self._provided_email = email
+        self._log = log
+
+    def __str__(self) -> str:
+        return self.uid
+
+    @cached_property
+    def email(self) -> str:
+        """Email parsed from the signing key"""
+        return parseaddr(self.signing_key["uid"])[1]
```
Comment (phlax) [inline tools/gpg/identity.py:37]: `self.uid`

File: tools/gpg/identity.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+    @property
+    def provided_name(self) -> Optional[str]:
+        """Provided name for the identity"""
+        return self._provided_name
+
+    @cached_property
+    def name(self) -> str:
+        """Name parsed from the signing key"""
+        return parseaddr(self.signing_key["uid"])[0]
```
Comment (phlax) [inline tools/gpg/identity.py:84]: `self.uid`

File: tools/testing/BUILD [hunk 6]
```diff
@@ -19,13 +19,13 @@ envoy_py_binary(
...
         "//:pytest.ini",
-        "//tools/base:runner",
-        "//tools/base:utils",
     ],
     deps = [
         requirement("pytest"),
         requirement("pytest-asyncio"),
         requirement("pytest-cov"),
+        "//tools/base:runner",
+        "//tools/base:utils",
```
Comment (phlax) [inline tools/testing/BUILD:28]: this ensures that depending on this rule you automatically depend on their dependencies

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**